### PR TITLE
hid: Add support for multiplayer and multilayout controllers

### PR DESCRIPTION
**Pull request by DarkLordZach**
The fabled "HID 2" successor to #1444. This would not have been possible without HOS legend @ogniK5377 for his backend work to make this possible!

- Adds full UI support to change the connectivity, type/layouts and buttons for all eight players, the handheld controller, the debug controller, mouse, keyboard, and finally touchscreen.
- Moves the docked mode setting to the input tab as it directly affects whether or not handheld can be enabled
- Automatically rotates the left and right joycons when used in a single-joycon configuration
- Will automatically change your configuration to the closest supported one if the game rejects it
- Avoids reloading the game list on every config change

here are some pics of the UI:
hid2-1
hid2-2
hid-2-3
hid-2-4

### CODE REVIEW

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 1]
```diff
@@ -433,7 +485,53 @@ Kernel::SharedPtr<Kernel::Event> Controller_NPad::GetStyleSetChangedEvent() cons
...
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+        return static_cast<std::size_t>(npad_id);
+    case 8:
+    case 32:
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/npad.cpp]: We should probably have a constexpr for these magic numbers to indicate they're handhelds

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 2]
```diff
@@ -433,7 +485,53 @@ Kernel::SharedPtr<Kernel::Event> Controller_NPad::GetStyleSetChangedEvent() cons
...
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+        return static_cast<u32>(index);
+    case 8:
+        return 32;
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/npad.cpp]: ditto

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 3]
```diff
@@ -451,6 +549,18 @@ void Controller_NPad::AddNewController(NPadControllerType controller) {
     InitNewlyAddedControler(controller_id);
 }
 
+void Controller_NPad::AddNewControllerAt(NPadControllerType controller, u32 npad_id) {
+    controller = DecideBestController(controller);
+    if (controller == NPadControllerType::Handheld) {
+        connected_controllers[8] = {controller, true};
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/npad.cpp]: ditto

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 4]
```diff
@@ -463,6 +573,122 @@ void Controller_NPad::DisconnectNPad(u32 npad_id) {
...
+        }
+    }
+    UNIMPLEMENTED_MSG("Could not find supported controller!");
+    return priority;
+}
+
+bool Controller_NPad::IsControllerSupported(NPadControllerType controller) {
+    if (controller == NPadControllerType::Handheld) {
+        // Handheld is not even a supported type, lets stop here
+        if (std::find(supported_npad_id_types.begin(), supported_npad_id_types.end(), 32) ==
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/npad.cpp]: ditto

File: src/core/hle/service/hid/controllers/touchscreen.cpp [hunk 5]
```diff
@@ -41,25 +41,32 @@ void Controller_Touchscreen::OnUpdate(u8* data, std::size_t size) {
 
     const auto [x, y, pressed] = touch_device->GetStatus();
     auto& touch_entry = cur_entry.states[0];
-    if (pressed) {
+    touch_entry.attribute.raw = 0;
+    if (pressed && Settings::values.touchscreen.enabled) {
+        if (cur_entry.entry_count == 0) {
+            touch_entry.attribute.start_touch.Assign(1);
```
Comment (ogniK5377) [inline src/core/hle/service/hid/controllers/touchscreen.cpp]: attributes should be removed since they're leftovers from research and are not actually tested

File: src/yuzu/configuration/configure_input.cpp [hunk 6]
```diff
@@ -9,334 +9,164 @@
...
 
-void ConfigureInput::restoreDefaults() {
-    for (int button_id = 0; button_id < Settings::NativeButton::NumButtons; button_id++) {
-        buttons_param[button_id] = Common::ParamPackage{
-            InputCommon::GenerateKeyboardParam(Config::default_buttons[button_id])};
-    }
+template <typename Dialog, typename... Args>
+void ConfigureInput::CallConfigureDialog(Args... args) {
+    this->applyConfiguration();
+    Dialog dialog(this, args...);
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input.cpp]:
```cpp
Dialog dialog{this, std::forward<Args>(args)...);

File: src/yuzu/configuration/configure_input.cpp [hunk 7]
```diff
@@ -9,334 +9,164 @@
...
+    ui->use_docked_mode->setEnabled(!Core::System::GetInstance().IsPoweredOn());
 }
 
-void ConfigureInput::restoreDefaults() {
-    for (int button_id = 0; button_id < Settings::NativeButton::NumButtons; button_id++) {
-        buttons_param[button_id] = Common::ParamPackage{
-            InputCommon::GenerateKeyboardParam(Config::default_buttons[button_id])};
-    }
+template <typename Dialog, typename... Args>
+void ConfigureInput::CallConfigureDialog(Args... args) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input.cpp]: `Args&&...`

File: src/yuzu/configuration/configure_input.cpp [hunk 8]
```diff
@@ -9,334 +9,164 @@
...
+    Settings::values.keyboard_enabled = ui->keyboard_enabled->isChecked();
+    Settings::values.touchscreen.enabled = ui->touchscreen_enabled->isChecked();
 }
 
-void ConfigureInput::updateButtonLabels() {
-    for (int button = 0; button < Settings::NativeButton::NumButtons; button++) {
-        button_map[button]->setText(ButtonToText(buttons_param[button]));
-    }
+void ConfigureInput::updateUIEnabled() {
+    for (std::size_t i = 0; i < 8; ++i) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input.cpp]: Magic value. (Ditto for all other occurrences of `8` below). These should be using the respective array's .size() member function instead of hardcoding the value.

File: src/yuzu/configuration/configure_input.cpp [hunk 9]
```diff
@@ -9,334 +9,164 @@
...
-    device_pollers = InputCommon::Polling::GetPollers(type);
-
-    // Keyboard keys can only be used as button devices
-    want_keyboard_keys = type == InputCommon::Polling::DeviceType::Button;
 
-    for (auto& poller : device_pollers) {
-        poller->Start();
+    bool hit_disabled = false;
+    for (auto* player : players_enabled) {
+        if (hit_disabled)
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input.cpp]: This can be collapsed as:

```cpp
player->setDisabled(hit_disabled);
```

File: src/yuzu/configuration/configure_input_player.cpp [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+static void LayerGridElements(QGridLayout* grid, QWidget* item, QWidget* onTopOf) {
+    const int index1 = grid->indexOf(item);
+    const int index2 = grid->indexOf(onTopOf);
+    int row, column, rowSpan, columnSpan;
+    grid->getItemPosition(index2, &row, &column, &rowSpan, &columnSpan);
+    grid->takeAt(index1);
+    grid->addWidget(item, row, column, rowSpan, columnSpan);
+}
+
+static QString getKeyName(int key_code) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input_player.cpp]: GetKeyName

File: src/yuzu/configuration/configure_mouse_advanced.cpp [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+#include <QMessageBox>
+#include <QTimer>
+#include "common/assert.h"
+#include "common/param_package.h"
+#include "input_common/main.h"
+#include "ui_configure_mouse_advanced.h"
+#include "yuzu/configuration/config.h"
+#include "yuzu/configuration/configure_mouse_advanced.h"
+
+static QString getKeyName(int key_code) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_mouse_advanced.cpp]: This is duplicating code that's already in configure_input_player.cpp. Consider moving both to a common spot.

File: src/yuzu/configuration/configure_mouse_advanced.cpp [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    case Qt::Key_Alt:
+        return QObject::tr("Alt");
+    case Qt::Key_Meta:
+        return "";
+    default:
+        return QKeySequence(key_code).toString();
+    }
+}
+
+static QString ButtonToText(const Common::ParamPackage& param) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_mouse_advanced.cpp]: Ditto

File: src/yuzu/configuration/configure_mouse_advanced.h [hunk 13]
```diff
@@ -0,0 +1,67 @@
+﻿// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <QDialog>
+#include <QWidget>
+#include <boost/optional.hpp>
```
Comment (lioncash) [inline src/yuzu/configuration/configure_mouse_advanced.h]: This should be using the standard optional header now.

File: src/yuzu/configuration/configure_input_player.h [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+#pragma once
+
+#include <array>
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <QDialog>
+#include <QKeyEvent>
+#include <boost/optional.hpp>
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input_player.h]: This should be using the standard optional header now.

File: src/yuzu/configuration/configure_touchscreen_advanced.cpp [hunk 15]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2016 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <algorithm>
+#include <memory>
```
Comment (lioncash) [inline src/yuzu/configuration/configure_touchscreen_advanced.cpp]: Many of these includes seem unnecessary in this file.

File: src/yuzu/configuration/configure_touchscreen_advanced.cpp [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+#include "yuzu/configuration/configure_touchscreen_advanced.h"
+
+ConfigureTouchscreenAdvanced::ConfigureTouchscreenAdvanced(QWidget* parent)
+    : QDialog(parent), ui(std::make_unique<Ui::ConfigureTouchscreenAdvanced>()) {
+    ui->setupUi(this);
+
+    connect(ui->restore_defaults_button, &QPushButton::pressed, this,
+            &ConfigureTouchscreenAdvanced::restoreDefaults);
+
+    this->loadConfiguration();
```
Comment (lioncash) [inline src/yuzu/configuration/configure_touchscreen_advanced.cpp]: `this->` isn't necessary here.

File: src/yuzu/configuration/configure_touchscreen_advanced.h [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Ui {
+class ConfigureTouchscreenAdvanced;
+}
+
+class ConfigureTouchscreenAdvanced : public QDialog {
+    Q_OBJECT
+
+public:
+    explicit ConfigureTouchscreenAdvanced(QWidget* parent);
```
Comment (lioncash) [inline src/yuzu/configuration/configure_touchscreen_advanced.h]: destructor should also be in the cpp file.

**Comment by ogniK5377** [comment]
A handheld controller cannot be connected unless all other npads are connected. A possible reason is that handheld controllers are detected as a generic npad controller which it shouldn't be. Besides that the npad priority switching lgtm.

**Comment by FearlessTobi** [comment]
Reminder to remove the checkboxes and to replace them with a none option, in case you haven't already done it.

File: src/yuzu/configuration/configure_gamelist.cpp [hunk 18]
```diff
@@ -36,6 +36,16 @@ ConfigureGameList::ConfigureGameList(QWidget* parent)
     InitializeRowComboBoxes();
 
     this->setConfiguration();
+
+    // Force game list reload if any of the relevant settings are changed.
```
Comment (FearlessTobi) [inline src/yuzu/configuration/configure_gamelist.cpp]: How is this commit related in any way to HID? If there is not a special reason for it being here, it should be dropped and put into its own PR. The diff is already big enough without stuff like this.

File: src/yuzu/main.cpp [hunk 19]
```diff
@@ -1374,6 +1381,11 @@ void GMainWindow::OnLoadAmiibo() {
     }
 }
 
+void GMainWindow::OnOpenYuzuFolder() {
+    QDesktopServices::openUrl(QUrl::fromLocalFile(
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp]: Ditto.

**Comment by ReaperOfSouls1909** [comment]
https://media.discordapp.net/attachments/472575665334648833/509924211796475904/unknown.png?width=1041&height=586

https://cdn.discordapp.com/attachments/472575665334648833/509922698382671872/unknown.png

https://cdn.discordapp.com/attachments/472575665334648833/509922334111563776/unknown.png 

Works well not found any issues yet going to test more games

**Comment by jroweboy** [comment]
Reminder to fix conflicts please ❤️

**Comment by FearlessTobi** [comment]
Some questions/nits about the UI:

- Please stretch out the `Configure Input` dialog

- https://cdn.discordapp.com/attachments/506213304897634308/511193566140432395/Unbenannt.PNG
  Which checkbox does the text here talk about?

- Why is it possible to disable the touchscreen?

- Is the touchscreen configuration and the debug controller something we should really put into the Input tab? Remember we design the UI mainly for end users and I'd love to avoid issues when they mess around with settings and cause their input not to work correctly anymore.

**Comment by bunnei** [comment]
Needs a rebase

**Comment by Hexagon12** [comment]
Reminder to rebase.
